### PR TITLE
move 2 packages (resource, shared) from evs-transcode-node to evs-com…

### PR DESCRIPTION
**Pull request by rmikhael**
…mon, to be used by evs-transcode-master.  

Jira: VOD-2179

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -10,8 +10,10 @@ This contains a common library for Ellation Video Services (EVS) VOD playback.
 * newrelic - new relic request handling integration.  In the future we plan to extend this to include application metrics beyond this.
 * objects - an object store abstraction.  Has support for http, s3, a static store mainly used for testing, and redis.  Redis wraps another store and
 effectively implements a simple write-back cache.
+* resource - package for S3 interactions
```
[the remark this answers is not in the document]
Comment (rmikhael) [reply README.md]: A Resource is a remote file location that may be uploaded to or downloaded from. The package implements 2 concrete types that satisfy the interface:  file:/// URLs  and  s3:/// URLs, but can be extended to other resources that can be downloaded from / uploaded to.
S3 is the main dish, but it is extensible.
I d rather keep the current name, cause it is used extensively in the node repo. We can add a techDebt for this if needed.

[the remark this answers is not in the document]
Comment (rmikhael) [reply README.md]: I can use a better comment above though. How about "* resource - package for S3/file uploads and downloads"?

File: resource/file.go [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+)
+
+const CopyChunkSize = 1024 * 1024 // 1MiB
+
+type file struct {
+	URL  string
+	Path string
+}
+
+// FileOpen handles file:/// URLs.
```
[the remark this answers is not in the document]
Comment (rmikhael) [reply resource/file.go]: No, I think this is trying to access files on mounted drives on the network?
something like what we access on Windows Explorer using "//sv-filer/backup/"

I am just guessing here.
But the S3 support is in file S3.go

File: resource/file.go [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+	}
+
+	if parsed.Path == "" {
+		return nil, fmt.Errorf("Couldn't parse file url %v: no path", u)
+	}
+
+	return file{u, parsed.Path}, nil
+}
+
+func (l file) Download(into string, cancel <-chan struct{}) error {
```
[the remark this answers is not in the document]
Comment (rmikhael) [reply resource/file.go]: Method signature has to follow what the interface has defined, or else it won't satisfy the interface.
So it is mandatory here.

What would close an open file during download, maybe OS or a machine/disk shut-down?

File: resource/file.go [hunk 4]
```diff
@@ -0,0 +1,93 @@
+package resource
+
+import (
+	"fmt"
+	"io"
+	"net/url"
+	"os"
+)
+
+const CopyChunkSize = 1024 * 1024 // 1MiB
```
[the remark this answers is not in the document]
Comment (mdale) [reply resource/file.go]: agree. Will this always be 1 MB ?

File: resource/retry.go [hunk 5]
```diff
@@ -0,0 +1,41 @@
+package resource
+
+import (
+	"time"
+)
+
+var retryInterval = time.Second * 5
```
Comment (mdale) [inline resource/retry.go]: also should probably be in config.

Comment (mdale) [reply resource/retry.go]: do we have a concept of giving up after a number of retries ?

File: resource/s3.go [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+func (l s3url) Download(into string, cancel <-chan struct{}) error {
+	file, err := os.Create(into)
+	if err != nil {
+		return RemoteError{"download", l.URL, err}
+	}
+	defer file.Close()
+
+	// AWS Role Credentials or static from local config
+	s := session.New()
+	s.Config.MaxRetries = aws.Int(3) // try a few times
```
Comment (mdale) [inline resource/s3.go]: MaxRetries could come from config instead of being defined inline here ?

**Comment by mdale** [comment]
* Please include a much more descriptive pull request summary. See https://github.com/crunchyroll/ellation_formation/blob/master/.github/PULL_REQUEST_TEMPLATE.md as a good template we should adopt for our projects. 
* Do we have a sense of new-relic integration and or logging levels for these download actions ( i.e how will we debug this on production when things go wrong ? ) how are we monitoring failure rates for s3 copies etc.  
* Is it possible to include unit-tests for these common utilities, or is that handled elsewhere ?
* some rational around retires counts and chunk sizes should be presented. 
* If this is a common library we should describe in the pull request what other projects will be using this utility, what it replaces etc.

File: resource/retry.go (hunk 5 above)
Comment (rmikhael) [reply resource/retry.go]: var Tries = 12
return RetryResource{res, Tries}, nil

both in file opener.go. So we retry opening the resource 12 times.

File: resource/s3.go (hunk 6 above)
Comment (rmikhael) [reply resource/s3.go]: Yes, we currently try 3 times for download, 12 times for upload. Both hard-coded.

To be configured along with many other constants requested by Subhrendu.

File: resource/file.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (rmikhael) [reply resource/file.go]: Yes, but only for file (I am not sure when that is exercised) not for S3 uploads/downloads. S3 uses a fixed 5 MByte chunks for up to 5 concurrent downloads.

**Comment by rmikhael** [comment]
To Michael's questions/comments:
- Will try to follow the template. This is just "copying" of the 2 directories from node up to evs-common, without much changes. So didn't try to optimize each file/function cause I am not adding them.
- The node and master use Statter + Notifier. I am not sure what you mean by new-relic.
- The node and master don't have any _test.go files. That is something we will have to review.
- Currently these packages are used only in the node, and I am moving them to common so the master can use them as well. No other services/repos are using them as far as I know.

**Pull request merged**